[14.0][MIG] l10n_it_fatturapa_in

TheMule71 opened the pull request:
Migrazione dalla 12.0 alla 14.0

DA COMPLETARE

--
Confermo di aver firmato il CLA https://odoo-community.org/page/cla e di aver letto le linee guida su https://odoo-community.org/page/contributing

SimoRubi commented:
> Migrazione dalla 12.0 alla 14.0
> 
> DA COMPLETARE
> 
> --
> Confermo di aver firmato il CLA https://odoo-community.org/page/cla e di aver letto le linee guida su https://odoo-community.org/page/contributing

@TheMule71 puoi fare delle PR in bozza e segnarle come 'pronte per review' quando le hai completate

TheMule71 commented:
@SimoRubi posso cambiare lo stato anche adesso?

SimoRubi commented:
> @SimoRubi posso cambiare lo stato anche adesso? o va rifatta?

Puoi farlo anche adesso, dovresti avere un 'convert to draft' vicino a dove sarà la lista dei reviewer:
![image](https://user-images.githubusercontent.com/32064796/102595748-671e6d80-4118-11eb-8759-5f823effcd8a.png)

TheMule71 commented:
Fatto, grazie.

dcorio commented:
@TheMule71 FYI: il modulo pare funzionare correttamente con xmlschema 1.4.x ma non con xmlschema 1.5.x.

Con 1.5 ricevo:
'dict' must be a MutableMapping subclass, not <class 'odoo.addons.l10n_it_fatturapa_in.wizard.efattura.CreateFromDocument.<locals>.ObjectDict'>

eLBati commented:
https://github.com/OCA/l10n-italy/pull/2174

As400it commented:
Un esempio di fattura in con cassa e ritenuta d'acconto che al momento rompe l'import. 


[IT01641790702.xml.zip](https://github.com/OCA/l10n-italy/files/6210838/IT01641790702.xml.zip)

dcorio commented on l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py:
```diff
@@ -0,0 +1,1556 @@
+            self.set_roundings(FatturaBody, invoice)
+
+        # compute the invoice
+        invoice._move_autocomplete_invoice_lines_values()
```
@TheMule71 
In caso di fatture con arrotondamento IVA quel _move_autocomplete_invoice_line_values pare che faccia casino.
Ad esempio con quasi tutte le fatture amazon. Ho provato a capirci qualcosa ma non ho risolto nulla.
Posso fornire un XML di esempio.

Borruso commented on l10n_it_fatturapa_in/models/attachment.py:
```diff
@@ -0,0 +1,139 @@
+            wiz_obj = self.env["wizard.import.fatturapa"].with_context(
+                from_attachment=att
+            )
+            if wiz_obj:
```
Dato che `wiz_obj` non ha dati, tutti i campi messi a `False` sopra non verranno mai popolati.
Potresti togliere la `if` così da far popolare gli altri campi.

eLBati commented:
Anche: https://github.com/OCA/l10n-italy/issues/2264

eLBati commented:
https://github.com/OCA/l10n-italy/issues/2273

eLBati commented:
https://github.com/OCA/l10n-italy/issues/2271

TheMule71 commented:
Se si mergia withholding_tax (e reasons) anche i test di fatturapa_in dovrebbero essere verdi:

`2021-05-14 14:50:26,274 35422 INFO odoo14-oca-wt3 odoo.modules.loading: Module l10n_it_fatturapa_in: loading demo 
2021-05-14 14:50:26,298 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestDuplicatedAttachment.test_duplicated_attachment ... 
2021-05-14 14:50:31,188 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAEnasarco.test_01_xml_import_enasarco ... 
2021-05-14 14:50:34,137 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAEnasarco.test_02_xml_import_enasarco ... 
2021-05-14 14:50:36,747 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAEnasarco.test_03_xml_import_enasarco ... 
2021-05-14 14:50:39,318 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_00_xml_import ... 
2021-05-14 14:50:41,904 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_01_xml_import ... 
2021-05-14 14:50:44,707 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_01_xml_link ... 
2021-05-14 14:50:48,952 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_01_xml_zero_quantity_line ... 
2021-05-14 14:50:51,522 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_02_xml_import ... 
2021-05-14 14:50:53,999 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_02_xml_link ... 
2021-05-14 14:50:58,285 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_03_xml_import ... 
2021-05-14 14:51:01,397 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_04_xml_import ... 
2021-05-14 14:51:04,276 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_05_xml_import ... 
2021-05-14 14:51:06,765 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_06_import_except ... 
2021-05-14 14:51:07,722 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_07_xml_import ... 
2021-05-14 14:51:10,291 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_08_xml_import ... 
2021-05-14 14:51:13,023 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_09_xml_import ... 
2021-05-14 14:51:15,681 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_10_xml_import ... 
2021-05-14 14:51:18,234 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_11_xml_import ... 
2021-05-14 14:51:20,858 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_12_xml_import ... 
2021-05-14 14:51:23,574 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_13_xml_import ... 
2021-05-14 14:51:27,762 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_14_xml_import ... 
2021-05-14 14:51:30,447 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_15_xml_import ... 
2021-05-14 14:51:33,005 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_16_xml_import ... 
2021-05-14 14:51:42,111 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_17_xml_import ... 
2021-05-14 14:51:44,838 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_19_xml_import ... 
2021-05-14 14:51:47,426 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_20_xml_import ... 
2021-05-14 14:51:50,021 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_21_xml_import ... 
2021-05-14 14:51:51,540 35422 INFO ? odoo.http: HTTP Configuring static files 
2021-05-14 14:51:53,145 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_22_xml_import ... 
2021-05-14 14:51:54,397 35422 WARNING odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.wizard.efattura: removed invalid dateTime element /p:FatturaElettronica/FatturaElettronicaBody/DatiGenerali/DatiTrasporto/DataOraConsegna: 0001-01-01T00:00:00.000+02:00 () 
2021-05-14 14:51:55,167 35422 WARNING odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.wizard.efattura: removed invalid dateTime element /p:FatturaElettronica/FatturaElettronicaBody/DatiGenerali/DatiTrasporto/DataOraConsegna: 0001-01-01T00:00:00.000+02:00 () 
2021-05-14 14:51:56,072 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_23_xml_import ... 
2021-05-14 14:51:59,455 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_24_xml_import ... 
2021-05-14 14:52:02,225 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_25_xml_import ... 
2021-05-14 14:52:05,160 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_26_xml_import ... 
2021-05-14 14:52:07,849 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_30_xml_import ... 
2021-05-14 14:52:13,138 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_31_xml_import ... 
2021-05-14 14:52:15,800 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_32_xml_import ... 
2021-05-14 14:52:18,565 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_33_xml_import ... 
2021-05-14 14:52:21,253 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_34_xml_import ... 
2021-05-14 14:52:23,908 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_35_xml_import ... 
2021-05-14 14:52:26,633 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_36_xml_import ... 
2021-05-14 14:52:29,360 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_37_xml_import_dates ... 
2021-05-14 14:52:32,356 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_38_xml_import_dates ... 
2021-05-14 14:52:35,609 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_40_xml_import_withholding ... 
2021-05-14 14:52:38,275 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_41_xml_import_withholding ... 
2021-05-14 14:52:41,076 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_42_xml_import_withholding ... 
2021-05-14 14:52:43,916 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_43_xml_import_withholding ... 
2021-05-14 14:52:46,923 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_44_xml_import ... 
2021-05-14 14:52:49,712 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_45_xml_many_zeros ... 
2021-05-14 14:52:52,308 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_47_xml_import ... 
2021-05-14 14:52:55,216 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_48_xml_import ... 
2021-05-14 14:52:57,996 35422 INFO odoo14-oca-wt3 odoo.addons.l10n_it_fatturapa_in.tests.test_import_fatturapa_xml: Starting TestFatturaPAXMLValidation.test_xml_import_summary_tax_rate ... 
2021-05-14 14:53:00,961 35422 INFO odoo14-oca-wt3 odoo.modules.loading: Module l10n_it_fatturapa_in loaded in 157.28s (incl. 154.65s test), 587 queries (+36048 test) 
`

eLBati commented on l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py:
```diff
@@ -998,7 +998,8 @@ def invoiceCreate(self, fatt, fatturapa_attachment, FatturaBody, partner_id):
 
-        self.set_roundings(FatturaBody, invoice)
+        if self.e_invoice_detail_level != "1":
+            self.set_roundings(FatturaBody, invoice)
```
Questo sarebbe da fare anche su v12?

TheMule71 replied:
Il problema me lo aveva segnalato  @mboscolo (e304917e9d3a7ca5ebddf353c6ae4ca2eca17eda) non ricordo se aveva provato anche sulla 12.0.

eLBati commented:
Anche https://github.com/OCA/l10n-italy/pull/2097

TheMule71 commented:
> Anche in `PrezzoTotale` e `ImponibileImporto`



> Come mai questo differisce dalla 12?
> 
> https://github.com/OCA/l10n-italy/blob/fabe988784c87b4b046d90e1009b8b72dcd8356c/l10n_it_fatturapa_in/tests/data/IT05979361218_016.xml#L58

Mi sa che di fatto sono due file diversi o quanto meno per test diversi. Di solito parto dall'ultimo test e ne copio il file aumentanto il progressivo, probabilmente ho fatto la stessa cosa in due momenti diversi sui due alberi 12 e 14.

Nell'attuale branch per la pr del porting alla 14, è stato introdotto a novembre con 7e9f3de89. Si tratta del file usato dal test_45_xml_many_zeros(). Probabilmente copiato da IT05979361218_015.xml e adattato.

Nella 12, è stato introdotto con eb67d3f0fab3d9da45eb2f41f55b0661568f7279, a febbraio (#2107), ed è il file per il test_46_xml_import(). Nella 12 è praticamente identico a IT05979361218_015.xml, per cui si potrebbe eliminarlo e modificare il test per usare IT05979361218_015.xml.

Il backporting del  test_45_xml_many_zeros() non è stato fatto (né ha senso farlo) perché è un problema specifico della 14.0 ("baco" di xmlschema/python).

L'alternativa è "bumpare" il numero del test sulla 14.0 (che va fatto comunque, perché test_45 sulla 12.0 fa un'altra cosa) e conseguentemente il nome del file (tipo IT05979361218_016.xml -> IT05979361218_017.xml).

Cosa di sembra meglio? Rimozione di un file mi pare inutile dalla 12.0 ("liberando" il nome IT05979361218_016.xml) oppure rename del test e del file sulla 14.0?

TheMule71 commented:
> Cosa di sembra meglio? Rimozione di un file mi pare inutile dalla 12.0 ("liberando" il nome IT05979361218_016.xml) oppure rename del test e del file sulla 14.0?

Anzi, ripensandoci, io farei così:
se non ci sono altre questioni aperte ovviamente, andrei verso il merge così com'è.

La PR è il porting di un branch parallelo della 12.0 (xmlschema) datato a novembre. Per es. il file IT05979361218_016.xml non c'era (quindi lo introduce da zero).

Immediatamente dopo il merge, si fa il forward porting delle PR sucessive sulla 12.0. Per es., #2000 introduce un test_45, per cui test_45_xml_many_zeros viene rinominato a test_46_xml_many_zeros come parte delle modifiche relative.

Quando #2107 verrà portata, (se decidiamo di mantenere il file extra sulla 12) il IT05979361218_016.xml della 14.0 verrà rinominato IT05979361218_017.xml come parte delle modifiche relative al porting.

In questo modo si ha l'history più pulita possibile.

È comunque un principio generale, nel fare il forward port delle modifiche queste cose succedono sempre, potenzialmente. Se una PR sulla 14 (e solo su quella) in futuro introdurrà per es. test_50_xxx  e poi successivamente una PR sulla 12 (ma per qualcosa che si applica anche alla 14) introdurrà test_50_yyy bisognerà gestire il caso di "collisione".

A meno che non decidiamo di far partire i numeri dei test per la sola 14 (da non backportare) da test_140 in su. In tal caso rinominiamo subito test_45_xml_many_zeros in test_140_xml_many_zeros e ce lo togliamo di torno anche per il futuro. Quello si può fare prima del merge oppure come prima modifica dopo (non cambia nulla).

TheMule71 commented:
Nel frattempo, ho squashato i commit come ci eravamo detti https://github.com/OCA/l10n-italy/pull/1985#issuecomment-846389401

Non c'è differenza tra il commit attuale e quello precedente alla push:
```
(odoo14) [marco@frodo (14.0-mig-l10n_it_fatturapa_in) l10n-italy.14.0-in]$ git diff d0828a8 d156798
(odoo14) [marco@frodo (14.0-mig-l10n_it_fatturapa_in) l10n-italy.14.0-in]$ 
```


Il rebase magari lo faccio dopo che sono stati mergiate #1985 e #1988.

TheMule71 commented:
Se volete pusho anche questo, oppure lo teniamo da parte per una PR successiva a merge. https://github.com/OCA/l10n-italy/commit/e586a23448974745aa9329db235bb141c1183a30

eLBati commented:
> Se volete pusho anche questo, oppure lo teniamo da parte per una PR successiva a merge. [e586a23](https://github.com/OCA/l10n-italy/commit/e586a23448974745aa9329db235bb141c1183a30)

Secondo me se ce l'hai già pronto funzionante conviene includerlo da subito

eLBati commented:
> Anche #2097

https://github.com/OCA/l10n-italy/issues/2302

SimoRubi commented on l10n_it_fatturapa_in/__manifest__.py:
```diff
@@ -0,0 +1,36 @@
+        "l10n_it_fatturapa",
+        "l10n_it_fiscal_document_type",
+        "l10n_it_withholding_tax_reason",
+        "l10n_it_fiscal_document_type",
```
Riga duplicata

Borruso commented on l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py:
```diff
@@ -0,0 +1,1722 @@
+            .xmlid_to_object("base.res_partner_rule", raise_if_not_found=False)
+        )
+        if vat:
+            domain = [("sanitized_vat", "=", vat)]
```
`sanitized_vat` alla v14 non esiste poiché lo aggiunge il modulo `base_vat_sanitized` in [OCA/partner-contact](https://github.com/OCA/partner-contact).
Inoltre questo modulo alla 14 è segnato come `obsoleto` ([Migration to version 14.0](https://github.com/OCA/partner-contact/issues/971))

TheMule71 commented:
Per chi fa la review: occhio a https://github.com/OCA/l10n-italy/pull/1990/commits/0aeaf3bbffd98c57d1c96d1d7c57eadbc05317fc#diff-2b5448b02e0e4d2ed90cc61085e73d322e6b7436bff0c08738af4e49f45c4bccR52-R58

È necessario per far funzionare il test45, che si base sulle vecchie funzionalità di base_vat_sanitized, in teoria incluse in base_vat, ma in pratica non fa le cose esattamente nello stesso modo.

Il test45 è stato modificato in questo modo:
- invece di settare partner_id.vat, faccio un write() perché è in quella parte del codice che vengono invocati i controlli di base_vat;
- il test non funzionava se invocato singolarmente (finalmente ho imparto come si fa!) perché prevedeva l'esistenza di un partner creato dal caricamento del file in un test precedente; adesso, se non c'è lo carica una prima volta, poi modifica il partner e prova a caricare il file una seconda volta.

Ho dovuto includere la fix proposta qui: https://github.com/odoo/odoo/pull/71920 perché il test45 si basa proprio sulla capacità di  rimuovere spazi dall'IVA anche in test. Piuttosto che modificare la logica del test (togliendo uno spazio in testa per farlo funzionare in un caso più favorevole) ho preferito includere la fix.

TheMule71 commented:
Vd. https://github.com/OCA/l10n-italy/issues/2321

> Versioni coinvolte:
> 14
> 
> Passi per riprodurre:
> Importare file xml da fattura fornitori
> 
> Comportamento osservato:
> Il campo "Numero documento" all'interno del file xml viene valorizzato all'interno del field payment_reference del modulo account.move anziché nel field ref (Ricevimento ricevuta)
> 
> Comportamento atteso:
> La valorizzazione del field ref (Ricevimento ricevuta) con il valore di "Numero documento" del file xml importato
> 
> Video/Screenshot (opzionale):
> ![Selection_119](https://user-images.githubusercontent.com/73834302/121375057-b82d5280-c940-11eb-8ee9-49471102fd81.png)

Nella 12.0 veniva copiato in reference, che è stato cancellato. Mi pare di averlo messo in ref all'inizio, poi si è optato di metterlo in payment_reference in base a questo commento: https://github.com/OCA/maintainer-tools/issues/430#issuecomment-633758952

Probabilmente nel nostro caso in effetti il campo ref è più corretto.

TheMule71 commented:
@AIR-sc 
Se ti va di provare questo:
```diff --git a/l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py b/l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py
index 9ed11959b..b2689227f 100644
--- a/l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py
+++ b/l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py
@@ -1091,7 +1091,7 @@ class WizardImportFatturapa(models.TransientModel):
                     ).date(),
                 }
             )
-        if not invoice.payment_reference:
+        if not invoice.ref:
             today = fields.Date.context_today(self)
             x = invoice.line_ids.filtered(
                 lambda line: line.account_id.user_type_id.type
@@ -1099,7 +1099,7 @@ class WizardImportFatturapa(models.TransientModel):
             ).sorted(lambda line: line.date_maturity or today)
             if x:
                 x[-1].name = FatturaBody.DatiGenerali.DatiGeneraliDocumento.Numero
-            invoice.payment_reference = (
+            invoice.ref = (
                 FatturaBody.DatiGenerali.DatiGeneraliDocumento.Numero
             )
```

TheMule71 commented:
Ho fatto un po' di squashing dei commit, facendo sparire i minifix. Rimangono:
- commit con contrbuti significativi (come quelli di @fredzamoabg);
- commit che sono forward-port espliciti di PR della 12.0;
- commit (come a0888fdb2) che potrebbero essere presi in considerazione per un backporting sulle versioni precedenti.

TheMule71 commented:
Ok, dopo varie peripezie (condite da pittoresche esclamazioni in Veneto e Aramaico Antico) ho stabilito che ref non è esattamente il posto migliore, visto che deve essere univoco. Anche nella 12.0 l'help del campo reference citava: "The payment communication that will be automatically populated once the invoice validation. You can also write a free communication."
https://github.com/odoo/odoo/blob/12.0/addons/account/models/account_invoice.py#L279

Quindi essendo testo libero non mi pare il caso di metterlo in un campo dichiarato univoco.

Il comportamento della 14 è coerente con quello della 12, usando payment_reference.

TheMule71 commented:
Riassunto puntate precendenti. Nel fare il tentativo di usare ref, ho dovuto modificare tutti i test per cancellare le fatture che caricavano sennò i test successivi fallivano tutti sul ref duplicato (negli XML i numeri sono più o meno sempre gli stessi).

Nel fare ciò, il test_08 si è messo, inspiegabilmente, a fallire (importo sbagliato). Curioso. Anche tornato alla situazione precedente, il test_08, se esiguito da solo, fallisce, se eseguito dopo test_07, funzia.

In pratica, se test_07 non esegue prima o non lascia i giro la sua fattura, test_08 non va. Sembra che funzioni solo grazie ad un side-effect del test_07, ma non sono riuscito ad identificarlo.

Non so quanto sia bloccante la cosa, ma sicuramente è strana/sospetta.

eLBati commented:
Nei test nella 12 il check `_check_duplicate_supplier_reference` non viene quasi mai eseguito perchè verrebbe eseguito alla conferma della fattura; conferma che nei test non viene quasi mai fatta.

Sulla 14, il `_check_duplicate_supplier_reference` esiste e ha lo stesso scopo, operando sul campo `ref`.

Quindi, ciò che nella 12 veniva scritto in `reference`, sulla 14 andrebbe scritto in `ref`, anche in base a come viene mostrato e descritto il campo.

Scrivere il `Numero` fattura in `payment_reference` mi sembra quindi sbagliato.

TheMule71 commented:
La descrizione del campo reference è "The **payment** communication that will be automatically populated once the invoice validation. You can also write a free communication"

TheMule71 commented:
Ultimo commit:
si copia il Numero sia in ref che payment_reference. I test sono stati modificati secondo il seguente criterio:
- se il test opera sul suo file, e nessun altro test lo usa, ho modificato direttamente il file, cambiando il Numero;
- se un file viene usato per test diversi, in tal caso modifico i test per cambiare ref e payment_reference alla fine del test o comunque dove oppurtuno. Questo è comunque inevitabile perché alcuni test caricano lo stesso file più volte (nello stesso test).

Se va bene lo affogo nel commit di Migrazione, prima del merge.

eLBati reviewed (approved):
Ok per me

eLBati commented:
/ocabot merge nobump

eLBati commented:
Approvato in chiamata
